packages/melody-component/src/hookComponent.js

```diff
@@ -0,0 +1,223 @@
+/**
+ * Copyright 2017 trivago N.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { enqueueComponent, options } from 'melody-idom';
+import {
+    shallowEquals,
+    shallowEqualsScalar,
+    shallowEqualsArray,
+} from './shallowEquals';
+
+let currentComponent = null;
+
+const HOOK_TYPE_USE_STATE = 1;
+const HOOK_TYPE_USE_EFFECT = 2;
+
+const enterHook = () => {
+    if (!currentComponent) {
+        throw new Error('Cannot use hooks outside of Component functions');
+    }
+    currentComponent.hooksPointer += 1;
+};
+
+export const useState = initialState => {
+    enterHook();
+    const { hooksPointer, hooks, state, setState } = currentComponent;
+
+    if (!currentComponent.isMounted) {
+        const setter = value => setState(hooksPointer, value);
+        const value = initialState;
+        hooks.push([HOOK_TYPE_USE_STATE, setter]);
+        state[hooksPointer] = value;
+        return [value, setter];
+    }
+    const setter = hooks[hooksPointer][1];
+    const value = state[hooksPointer];
+    return [value, setter];
+};
+
+export const useEffect = (callback, shouldUpdateOrDataArray) => {
```

Comment:
What is `shouldUpdateOrDataArray`?

Reply:
`shouldUpdateOrWatchArray` might be more understandable🤓

Reply:
Yes that's probably a better name. I also implemented an API slightly different from react. Not sure yet if it is a good idea:

```js
// Only runs on mount and on unmount
useEffect(() => { /* ... */ });

// Runs on mount, on every render & on unmount
useEffect(() => { /* ... */ }, true);

// Runs on mount, when at least one value has changed & on unmount
useEffect(() => { /* ... */ }, [foo, bar])
```

Reply:
React's API is likely to change the second parameter too, I think a lot of people didn't understand the empty array or forgot to use it

Reply:
ah cool :) do you have a link of the discussion?

Reply:
Little discussion here, nothing confirmed though:
https://twitter.com/dan_abramov/status/1057693582528835584

Reply:
cool, thanks!